ivadomed/ms-challenge-2021
# Major code refactoring using MONAI and PyTorch Lightning

This PR brings a major refactoring of the existing codebase by using Monai and Lightning. This is to improve the data loading aspects and provide more models for training.  

Important changes:

1. The structure of the data loader is now based on the Decathlon dataset i.e. in `utils/` there's a `create_data_json.py` which creates a json file that's according to the Decathlon dataset structure. This is done following MONAI's tutorials and it provides a function to load the dataset using `load_decathlon_datalist` given the json files as the input. Note the folder is still in BIDS format, only how the files are loaded and paired with each other is now different. 
    1. `utils/` now also contains `fold_generator.py` which is used for creating multiple datasets for k-fold cross validation. This is an useful addition for this and future projects!
3. The `nets/` folder currently contains `models.py` from the previous version of the codebase. The idea is to have 1 file per model in this folder in the future so that `models.py` does not become 10k lines long.
4. Lastly, `main_pl.py` uses transformations and pre-processing stuff from MONAI, DL-based stuff from PyTorch Lightning and DL-tracking stuff from WandB. The hope is to make the code easily understandable and reproducible with these modifications. 

Also, note that there are TODOs interspersed in `main_pl.py` which need some more time investigating and more importantly some discussion. These will be updated soon. 

The PR can be tested as it is, the training and validation parts work fine, testing also works (but needs some more modifications). Moreover, I was not planning on running testing code unless we have a good validation score, hence this PR can be considered complete! 

To test the PR, run the following commands:
1. `python create_json_data.py -se 30 -ncv 5 -dr ~/duke/temp/muena/ms-challenge-2021_preprocessed/data_processed_clean/` to create a 5-fold CV dataset. This should create 5 json files, like `dataset_fold-X.json` which contain the subjects for training/testing/validation. 
2. `python main_pl.py <your-arguments-here>`
